File: tools/clod-poc/src/simplify.ts
// SOLE meshoptimizer boundary. Plan §3.2 / §11.5.
//
// No other module imports meshoptimizer. quadtree.ts sees only this API.
// simplify_sloppy is NEVER used (topology-breaking; the plan forbids it outright).

import { MeshoptSimplifier } from "meshoptimizer";
import { PageMesh, ClodBuildError, vertexCount } from "./types.js";
import { ClodPagesConfig } from "./config.js";
import { paintMaterialAt } from "./terrain.js";

let ready = false;
export async function initSimplifier(): Promise<void> {
  if (ready) return;
  await MeshoptSimplifier.ready;
  // simplifyWithAttributes is gated behind this flag in some builds; harmless otherwise.
  (MeshoptSimplifier as unknown as { useExperimentalFeatures?: boolean }).useExperimentalFeatures = true;
  ready = true;
}

/** World-space simplification error scale for a mesh (meshopt_simplifyScale). */
export function simplifyScale(mesh: PageMesh): number {
  return MeshoptSimplifier.getScale(mesh.positions, 3);
}

export interface SimplifyOutput {
  mesh: PageMesh;
  resultError: number; // meshopt relative
  errorWorld: number; // resultError * simplifyScale
  lowBenefit: boolean;
}

/**
 * Decimate `mesh` toward target_ratio_per_level, carrying normals + material weights and
 * honouring per-vertex locks. Returns the simplified mesh plus world-space error.
 */
export function simplifyPage(
  mesh: PageMesh,
  locks: Uint8Array,
  cfg: ClodPagesConfig,
): SimplifyOutput {
  if (!ready) throw new ClodBuildError("SimplifierApiUnavailable", "call initSimplifier() first");

  const vc = vertexCount(mesh);
  const inputIndices = mesh.indices.length;
  const targetIndices = Math.max(3, Math.floor(inputIndices * cfg.simplify.target_ratio_per_level));

  // Interleave attributes: [n0 n1 n2 m0 m1 m2 m3] per vertex, stride 7.
  const ATTR_STRIDE = 7;
  const attrs = new Float32Array(vc * ATTR_STRIDE);
  for (let i = 0; i < vc; i++) {
    attrs[i * ATTR_STRIDE + 0] = mesh.normals[i * 3 + 0];
    attrs[i * ATTR_STRIDE + 1] = mesh.normals[i * 3 + 1];
    attrs[i * ATTR_STRIDE + 2] = mesh.normals[i * 3 + 2];
    attrs[i * ATTR_STRIDE + 3] = mesh.materials[i * 4 + 0];
    attrs[i * ATTR_STRIDE + 4] = mesh.materials[i * 4 + 1];
    attrs[i * ATTR_STRIDE + 5] = mesh.materials[i * 4 + 2];
    attrs[i * ATTR_STRIDE + 6] = mesh.materials[i * 4 + 3];
  }
  const wn = cfg.simplify.attribute_weights.normal;
  const wm = cfg.simplify.attribute_weights.material;
  const attrWeights = [wn, wn, wn, wm, wm, wm, wm];

  let result: [Uint32Array, number];
  try {
    result = MeshoptSimplifier.simplifyWithAttributes(
      mesh.indices,
      mesh.positions,
      3,
      attrs,
      ATTR_STRIDE,
      attrWeights,
      locks,
      targetIndices,
      cfg.simplify.target_error,
      ["LockBorder"],
    );
  } catch (e) {
    throw new ClodBuildError("MeshoptFailed", String(e));
  }

  const [newIndices, resultError] = result;

  // meshopt keeps the original vertex buffer; unused vertices are simply unreferenced.
  // Compact to referenced vertices so downstream weld/lock/stats stay tight.
  const compacted = compact(mesh, newIndices, cfg.simplify.weld_epsilon_cells);

  const errorWorld = resultError * simplifyScale(mesh);
  const lowBenefit = newIndices.length > cfg.simplify.abandon_ratio * inputIndices;

  return { mesh: compacted, resultError, errorWorld, lowBenefit };
}

/** Drop unreferenced vertices and remap indices. */
function snap(value: number, epsilon: number): number {
  return Math.round(value / epsilon) * epsilon;
}

function compact(mesh: PageMesh, indices: Uint32Array, snapEpsilon: number): PageMesh {
  const remap = new Map<number, number>();
  const pos: number[] = [], nrm: number[] = [], mat: number[] = [];
  const out = new Uint32Array(indices.length);
  for (let i = 0; i < indices.length; i++) {
    const old = indices[i];
    let ni = remap.get(old);
    if (ni === undefined) {
      ni = pos.length / 3;
      remap.set(old, ni);
      const px = snap(mesh.positions[old * 3], snapEpsilon);
      const py = snap(mesh.positions[old * 3 + 1], snapEpsilon);
      const pz = snap(mesh.positions[old * 3 + 2], snapEpsilon);
      pos.push(px, py, pz);
      nrm.push(mesh.normals[old * 3], mesh.normals[old * 3 + 1], mesh.normals[old * 3 + 2]);
      const [m0, m1, m2, m3] = paintMaterialAt(px, py, pz);
      mat.push(m0, m1, m2, m3);
    }
    out[i] = ni;
  }
  return {
    positions: new Float32Array(pos),
    normals: new Float32Array(nrm),
    materials: new Float32Array(mat),
    indices: out,
  };
}

File: tools/clod-rs/src/simplify.rs
//! SOLE meshoptimizer boundary (§11.5). Port of simplify.ts. Never simplify_sloppy.
//!
//! Carries normals + material weights and honours per-vertex locks. The attribute stride
//! passed to meshopt is in BYTES (PoC/Rust finding — the JS npm wrapper uses floats).

use crate::config::ClodPagesConfig;
use crate::types::PageMesh;
use meshopt::{simplify_scale, simplify_with_attributes_and_locks, SimplifyOptions, VertexDataAdapter};
use std::collections::HashMap;

const ATTR_STRIDE: usize = 7; // normal(3) + material(4)

pub struct SimplifyOutput {
    pub mesh: PageMesh,
    pub result_error: f32,
    pub error_world: f32,
    pub low_benefit: bool,
}

fn adapter(positions: &[[f32; 3]]) -> VertexDataAdapter<'_> {
    let bytes: &[u8] = bytemuck::cast_slice(positions);
    VertexDataAdapter::new(bytes, std::mem::size_of::<[f32; 3]>(), 0).expect("vertex adapter")
}

/// World-space simplification error scale for a mesh (meshopt_simplifyScale).
pub fn simplify_scale_of(mesh: &PageMesh) -> f32 {
    simplify_scale(&adapter(&mesh.positions))
}

pub fn simplify_page(mesh: &PageMesh, locks: &[bool], cfg: &ClodPagesConfig) -> SimplifyOutput {
    let vc = mesh.vertex_count();
    let input_indices = mesh.indices.len();
    let target = ((input_indices as f32 * cfg.simplify.target_ratio_per_level) as usize).max(3);

    let mut attrs = vec![0f32; vc * ATTR_STRIDE];
    for i in 0..vc {
        attrs[i * ATTR_STRIDE..i * ATTR_STRIDE + 3].copy_from_slice(&mesh.normals[i]);
        attrs[i * ATTR_STRIDE + 3..i * ATTR_STRIDE + 7].copy_from_slice(&mesh.materials[i]);
    }
    let wn = cfg.simplify.attribute_weights.normal;
    let wm = cfg.simplify.attribute_weights.material;
    let weights = [wn, wn, wn, wm, wm, wm, wm];

    let scale = simplify_scale_of(mesh);
    let mut result_error = 0f32;
    let new_indices = simplify_with_attributes_and_locks(
        &mesh.indices,
        &adapter(&mesh.positions),
        &attrs,
        &weights,
        ATTR_STRIDE * std::mem::size_of::<f32>(), // BYTES, not floats
        locks,
        target,
        cfg.simplify.target_error,
        SimplifyOptions::LockBorder,
        Some(&mut result_error),
    );

    let mesh = compact(mesh, &new_indices);
    let error_world = result_error * scale;
    let low_benefit = new_indices.len() as f32 > cfg.simplify.abandon_ratio * input_indices as f32;
    SimplifyOutput {
        mesh,
        result_error,
        error_world,
        low_benefit,
    }
}

/// Drop unreferenced vertices and remap indices.
fn compact(mesh: &PageMesh, indices: &[u32]) -> PageMesh {
    let mut remap: HashMap<u32, u32> = HashMap::new();
    let mut out = PageMesh::default();
    out.indices.reserve(indices.len());
    for &old in indices {
        let ni = *remap.entry(old).or_insert_with(|| {
            let ni = out.positions.len() as u32;
            out.positions.push(mesh.positions[old as usize]);
            out.normals.push(mesh.normals[old as usize]);
            out.materials.push(mesh.materials[old as usize]);
            ni
        });
        out.indices.push(ni);
    }
    out
}
